pageserver/src/page_service.rs

```diff
@@ -754,49 +730,58 @@ impl PageServerHandler {
             std::thread::sleep(std::time::Duration::from_millis(1000));
         }
         */
+
+        // FIXME: this profiling now happens at different place than it used to. The
+        // current profiling is based on a thread-local variable, so it doesn't work
+        // across awaits
+        let _profiling_guard = profpoint_start(self.conf, ProfilingConfig::PageRequests);
         let page = timeline.get_rel_page_at_lsn(req.rel, req.blkno, lsn)?;
 
         Ok(PagestreamBeMessage::GetPage(PagestreamGetPageResponse {
             page,
         }))
     }
 
-    fn handle_basebackup_request(
+    #[instrument(skip(self, pgb))]
+    async fn handle_basebackup_request(
         &self,
         pgb: &mut PostgresBackend,
+        tenant_id: ZTenantId,
         timeline_id: ZTimelineId,
         lsn: Option<Lsn>,
         prev_lsn: Option<Lsn>,
-        tenant_id: ZTenantId,
         full_backup: bool,
     ) -> anyhow::Result<()> {
-        let span = info_span!("basebackup", timeline = %timeline_id, tenant = %tenant_id, lsn = field::Empty);
-        let _enter = span.enter();
-        info!("starting");
-
         // check that the timeline exists
         let timeline = get_local_timeline(tenant_id, timeline_id)?;
         let latest_gc_cutoff_lsn = timeline.get_latest_gc_cutoff_lsn();
         if let Some(lsn) = lsn {
+            // Backup was requested at a particular LSN. Wait for it to arrive.
+            info!("waiting for {}", lsn);
+            timeline.wait_lsn(lsn).await?;
             timeline
                 .check_lsn_is_in_scope(lsn, &latest_gc_cutoff_lsn)
                 .context("invalid basebackup lsn")?;
         }
 
         // switch client to COPYOUT
         pgb.write_message(&BeMessage::CopyOutResponse)?;
+        pgb.flush().await?;
 
         /* Send a tarball of the latest layer on the timeline */
-        {
-            let mut writer = CopyDataSink { pgb };
-
+        let mut writer = CopyDataSink {
+            pgb,
+            rt: tokio::runtime::Handle::current(),
+        };
+        tokio::task::block_in_place(|| {
```

Comment:
Why would we use that instead of a `spawn_blocking` alternative?

I think, `block_in_place` is more dangerous, since its attempts to move other scheduled tasks to other workers might not work if our futures are not explicitly `Send + Sync`, while `spawn_blocking` moves away the blocking future instead, requiring it to be movable (`'static + Send + Sync`) which is better in terms of not-blocking the neighbour tasks.

Reply:
> Why would we use that instead of a spawn_blocking alternative?

With `spawn_blocking`, I get this:
```
error[E0759]: `pgb` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
   --> pageserver/src/page_service.rs:748:9
    |
748 |         pgb: &mut PostgresBackend,
    |         ^^^  -------------------- this data with an anonymous lifetime `'_`...
    |         |
    |         ...is used here...
...
776 |         tokio::task::spawn_blocking(|| {
    |         --------------------------- ...and is required to live as long as `'static` here
    |
note: `'static` lifetime requirement introduced by this bound
   --> /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/task/blocking.rs:194:35
    |
194 |         F: FnOnce() -> R + Send + 'static,
    |                                   ^^^^^^^

```
I couldn't figure out how to make that work, so I just went with `block_in_place` instead.

> I think, block_in_place is more dangerous, since its attempts to move other scheduled tasks to other workers might not work if our futures are not explicitly Send + Sync, while spawn_blocking moves away the blocking future instead, requiring it to be movable ('static + Send + Sync) which is better in terms of not-blocking the neighbour tasks.

Hmm, not sure what danger you see here. All tasks must be Send, unless you're using LocalSet. I don't see any discussion on how LocalSet and block_in_place work together in the tokio docs. I presume it wouldn't work.

I don't think Sync makes any difference here.

The danger I see and is mentioned in the docs for `block_in_place` is that if there are other Futures executing in the same Task, they will not run until this finishes. In our model, there's one Task per connection, and when we make this call, the connection will be busy with the basebackup and not doing anything else. So it seems OK to me. Launching the basebackup operation isn't performance critical either, so the effect to neighbouring tasks doesn't seem significant